ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -222,63 +222,14 @@ instance (ByronGiven, Typeable cfg, ConfigContainsGenesis cfg)
-                   (ByronLedgerState state snapshots) = do
-      runReaderT
-        (CC.Block.headerIsValid
-          (CC.Block.cvsUpdateState state)
-          (CC.Block.blockHeader block)
-        )
-        (fromBlockValidationMode CC.Block.BlockValidation)
-      CC.Block.BodyState { CC.Block.utxo, CC.Block.updateState
-                         , CC.Block.delegationState }
-        <- runReaderT
-            (CC.Block.updateBody bodyEnv bodyState block)
-            (fromBlockValidationMode CC.Block.BlockValidation)
-      let state' = state
-            { CC.Block.cvsLastSlot        = CC.Block.blockSlot block
-            , CC.Block.cvsPreviousHash    = Right $ CC.Block.blockHashAnnotated block
-            , CC.Block.cvsUtxo            = utxo
-            , CC.Block.cvsUpdateState     = updateState
-            , CC.Block.cvsDelegationState = delegationState
-            }
-          snapshots'
-              | CC.Block.cvsDelegationState state' ==
-                 CC.Block.cvsDelegationState state
-              = snapshots
-              | otherwise
-              = snapshots Seq.|> SB.bounded startOfSnapshot slot state'
-            where
-              startOfSnapshot = case snapshots of
-                _ Seq.:|> a -> sbUpper a
-                Seq.Empty   -> SlotNo 0
-              slot = convertSlot $ CC.Block.blockSlot block
-      return $ ByronLedgerState state' (trimSnapshots snapshots')
-    where
-      bodyState = CC.Block.BodyState
-        { CC.Block.utxo            = CC.Block.cvsUtxo state
-        , CC.Block.updateState     = CC.Block.cvsUpdateState state
-        , CC.Block.delegationState = CC.Block.cvsDelegationState state
-        }
-      bodyEnv = CC.Block.BodyEnvironment
-        { CC.Block.protocolMagic      = fixPM $ CC.Genesis.configProtocolMagic cfg
-        , CC.Block.k                  = CC.Genesis.configK cfg
-        , CC.Block.allowedDelegators  = allowedDelegators cfg
-        , CC.Block.protocolParameters = protocolParameters
-        , CC.Block.currentEpoch       = CC.Slot.slotNumberEpoch
-                                          (CC.Genesis.configEpochSlots cfg)
-                                          (CC.Block.blockSlot block)
-        }
-
-      protocolParameters = CC.UPI.adoptedProtocolParameters . CC.Block.cvsUpdateState
-                         $ state
-
-      fixPM (Crypto.AProtocolMagic a b) = Crypto.AProtocolMagic (reAnnotate a) b
-
-      k = CC.Genesis.configK cfg
+  applyLedgerBlock = applyByronLedgerBlock
+    (fromBlockValidationMode CC.Block.BlockValidation)
 
-      trimSnapshots = Seq.dropWhileL $ \ss ->
-        sbUpper ss < convertSlot (CC.Block.blockSlot block) - 2 * coerce k
+  reapplyLedgerBlock cfg blk st =
+    let validationMode = fromBlockValidationMode CC.Block.NoBlockValidation
+    in case runExcept (applyByronLedgerBlock validationMode cfg blk st) of
+      Left  err -> error ("reapplyLedgerBlock: unexpected error: " <> show err)
+      Right st' -> st'
```

Comment:
Ahh, so the type says it can still fail, because that's just how the ledger functions work. Ok, and that explains the `HasCallStack` above. Fair enough.

It's perhaps worth putting in a comment explaining what is going on here. ie that this shouldn't fail now because of the `NoBlockValidation` means the ledger code will not be doing any checks, but the type of `applyByronLedgerBlock` doesn't change based on the validation mode so it looks like it can still fail.

Reply:
Yes, this is exactly correct. Will doc this.

Reply:
Added some doc. Hopefully it's clear now.